Rust/c3id: Decrease allocations & formatting

jdraaijer-microsoft opened the pull request:
Reducing the amount of required allocations improves performance by about 27% for few-block computations (0 to 88 characters), and by about 7% for many block (1024 characters) due to reduced (re-)allocation overhead.

[Benchmark](https://github.com/microsoft/security-utilities/blob/users/jdraaijer/bench/src/security_utilities_rust/benches/cross_company_correlating_id.rs)

jdraaijer-microsoft committed:
fix(cross_company_correlating_id): Decrease allocations & formatting calls during generation of c3id

There are some steps in the generation of the Cross Company Correlation ID
for which we know the amount and type of data that we will be processing at
compile time. Passing this knowledge directly to those functions decreases
the amount of time spent formatting & reallocating at runtime.

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/cross_company_correlating_id.rs:
```diff
@@ -13,23 +13,25 @@ thread_local! {
-    fmt::format(format_args!("{:X}", result))
-}
\ No newline at end of file
+    let mut output = String::with_capacity(2 * 32);
```
@jdraaijer-microsoft  What's the rationale behind allocating 64 bytes? A rough upper bound? Can we store in a variable rather than hard code the value?

jdraaijer-microsoft replied:
We know this value exactly because hex encoding `n` bytes will always produce `n * 2` ascii bytes. 

Now that you point it out, I actually realised we don't need this allocation either: 
1. We know all the size up front.
2. The encoding is super simple and easy to verify.

Will add a commit that removes this one too, since we're hunting for % anyways.

jdraaijer-microsoft replied:
Done, this should now be ready to merge :)

jdraaijer-microsoft committed:
fix(cross_company_correlating_id): remove final allocation

Thanks to a comment by @suvamM I realized that there is no need to
allocate the string data at all. Since we know the length of the data
up front, and the formatting we do is so basic that doing it by hand
works fine, we can remove the allocation and gain another 7% speedup.

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/cross_company_correlating_id.rs:
```diff
@@ -2,34 +2,55 @@
-
+fn generate_sha256_hash<'a>(text: &str, buffer: &'a mut [u8; 64]) -> &'a [u8; 64] {
     let result = THREAD_LOCAL_SHA256.with(|sha| {
         sha.borrow_mut().update(text.as_bytes());
```
@jdraaijer-microsoft This can optimized as well right?

jdraaijer-microsoft replied:
If it's the base64 encoding you are referring to: yes, in theory. I'm less hopeful of doing that correctly though, so I think leaving it like this is OK. Re-implementing base64 encoding was not on my bucket list, unless we think it is worthwhile.

The main reason I don't think it is useful is because we need to return a heap-allocated string anyways (and the single allocation is the only thing we could improve). `STANDARD.encode` already optimizes the allocation for the string, so we cannot really improve there.

suvamM replied:
@jdraaijer-microsoft Referring to the double borrow happening in lines 30 and 31.

jdraaijer-microsoft replied:
Fixed

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/cross_company_correlating_id.rs:
```diff
@@ -2,34 +2,55 @@
\ No newline at end of file
+    let data: [u8; 32] = result.into();
+
+    for (idx, byte) in data.into_iter().enumerate() {
```
@jdraaijer-microsoft I don't entirely follow the algorithm here. Since the tests pass, I have faith in the logic. But a question: generating cross company ID should be relatively infrequent. Are the gains with this optimization high enough to trade off with respect to the readability?

jdraaijer-microsoft replied:
It saves one allocation (~7% perf for few-block hashes). This is just a manual implementation of UpperHex formatting for the array.

I can remove it or add a clarifying comment

jdraaijer-microsoft replied:
I've added a comment explaining what this does. IMO it's worth it since the encoding is so simple, but I can agree that it is a minor nuisance to understand

jdraaijer-microsoft committed:
fix(cross_company_correlating_id): borrow only once

jdraaijer-microsoft committed:
fix(cross_company_correlating_id): document hex encoding
